src/components/registration-form.js

```diff
@@ -1,148 +1,285 @@
+        </div>
       </div>
-    </div>
+    );
+  };
+
+  const renderEnterCodeForm = () => {
+    const verificationCodeValidationRules = [
+      {
+        required: true,
+        message: t('enterVerificationCode'),
+        whitespace: true
+      },
+      {
+        validator: (_rule, value) => {
+          return value && invalidVerificationCodes.includes(value)
+            ? Promise.reject(new Error(t('verificationCodeIsInvalid')))
+            : Promise.resolve();
+        }
+      }
+    ];
+    return (
+      <div className="RegistrationForm-form">
+        <Markdown className="RegistrationForm-verificationMessage">
+          {t('verificationMessage', { validityInMinutes: formatDuration(PENDING_USER_REGISTRATION_EXPIRATION_IN_MINUTES, 'minutes') })}
+        </Markdown>
+        <Form
+          form={enterCodeForm}
+          layout="vertical"
+          onFinish={handleEnterCodeFinish}
+          validateTrigger="onSubmit"
+          scrollToFirstError
+          >
+          <FormItem name="verificationCode" label={t('verificationCode')} rules={verificationCodeValidationRules}>
+            <Input />
+          </FormItem>
+        </Form>
+        <div className="RegistrationForm-registerButton">
+          <Button type="primary" size="large" onClick={handleEnterCodeButtonClick} block>
+            {t('common:confirm')}
+          </Button>
+        </div>
+      </div>
+    );
+  };
+
+  const renderCompletedMessage = () => (
+    <Result
+      status="success"
+      title={t('completedMessageTitle')}
+      subTitle={t('completedMessageSubTitle')}
+      extra={[<Button key="login" type="primary" onClick={handleGotoDashboardClick}>{t('gotoDashboard')}</Button>]}
+      />
   );
 
-  const registrationConfirmation = (
-    <Markdown className="RegistrationForm-confirmation">
-      {t('registrationConfirmationMessage', { loginPageUrl: routes.getLoginUrl() })}
-    </Markdown>
+  const renderStepTitel = step => (
+    <span className={classNames({ 'RegistrationForm-step': true, 'is-selected': currentStepKey === step.key })}>
+      {t(step.translationKey)}
+    </span>
   );
 
+  const renderStepIcon = step => isLoading && currentStepKey === step.key ? <LoadingOutlined /> : step.icon;
+
+  const renderSteps = () => {
+    const stepItems = steps.map(step => ({
+      title: renderStepTitel(step),
+      icon: renderStepIcon(step),
+      status: step.status
+    }));
+    return (
+      <div className="RegistrationForm-steps">
+        <Steps items={stepItems} labelPlacement="vertical" size="small" />
+      </div>
+    );
+  };
+
   return (
-    <div className="RegistrationForm">
-      {user ? registrationConfirmation : registrationForm}
+    <div className="RegistrationForm u-panel">
+      {renderSteps()}
+      {currentStepKey === STEP.enterData && renderEnterDataForm()}
+      {currentStepKey === STEP.enterCode && renderEnterCodeForm()}
+      {currentStepKey === STEP.completed && renderCompletedMessage()}
```

Comment:
Nice switching mechanism with these steps and states! 👍

Reply:
:+1: